ml-agents/mlagents/trainers/trainer.py

```diff
@@ -165,97 +189,129 @@ def export_model(self, name_behavior_id: str) -> None:
+    @abc.abstractmethod
+    def end_episode(self):
         """
-        Uses demonstration_buffer to update model.
+        A signal that the Episode has ended. The buffer must be reset.
+        Get only called when the academy resets.
         """
-        raise UnityTrainerException("The update_model method was not implemented.")
+        pass
 
+    @abc.abstractmethod
     def create_policy(self, brain_parameters: BrainParameters) -> TFPolicy:
         """
         Creates policy
         """
-        raise UnityTrainerException("The create_policy method was not implemented.")
+        return None
 
+    @abc.abstractmethod
     def add_policy(self, name_behavior_id: str, policy: TFPolicy) -> None:
         """
         Adds policy to trainer
         """
-        raise UnityTrainerException("The add_policy method was not implemented")
+        pass
 
+    @abc.abstractmethod
     def get_policy(self, name_behavior_id: str) -> TFPolicy:
         """
         Gets policy from trainer
         """
-        raise UnityTrainerException("The get_policy method was not implemented.")
+        return None
 
-    def advance(self) -> None:
+    @abc.abstractmethod
+    def _is_ready_update(self):
+        """
+        Returns whether or not the trainer has enough elements to run update model
+        :return: A boolean corresponding to wether or not update_model() can be run
+        """
+        return False
+
+    @abc.abstractmethod
+    def _update_policy(self):
+        """
+        Uses demonstration_buffer to update model.
+        """
         pass
+
+    def advance(self) -> None:
+        """
+        Steps the trainer, taking in trajectories and updates if ready.
+        """
+        with hierarchical_timer("process_trajectory"):
+            for traj_queue in self.trajectory_queues:
+                if not traj_queue.empty():
+                    _t = traj_queue.get_nowait()
+                    self._process_trajectory(_t)
+        if self.should_still_train:
+            if self._is_ready_update():
+                with hierarchical_timer("_update_policy"):
+                    self._update_policy()
+                    for q in self.policy_queues:
+                        # Get policies that correspond to the policy queue in question
+                        q.put(self.get_policy(q.behavior_id))
+
+    def publish_policy_queue(self, policy_queue: Queue) -> None:
+        """
+        Adds a policy queue to the list of queues to publish to when this Trainer
+        makes a policy update
+        :param queue: Policy queue to publish to.
+        """
+        self.policy_queues.append(policy_queue)
+
+    def subscribe_trajectory_queue(self, trajectory_queue: Queue) -> None:
+        """
+        Adds a trajectory queue to the list of queues for the trainer injest Trajectories from.
+        :param queue: Trajectory queue to publish to.
+        """
+        self.trajectory_queues.append(trajectory_queue)
```

Comment:
Wrong types here - without https://github.com/Unity-Technologies/ml-agents/pull/3151 mypy doesn't understand what an AgentManagerQueue is. With it, you'll get an error like
```
Argument 1 to "append" of "list" has incompatible type "Queue[Any]"; expected "AgentManagerQueue"
```